### DISCUSSION TITLE
Preroll Plus not finding Plex server

### DISCUSSION THREAD

**Discussion by imsorrywhat** [chadwpalm/PrerollPlus#1] [Q&A]
When clicking the circle arrows to find Plex server, nothing is returned and the container logs show an "Invalid URL" and 401 error.

Overseerr, running on same Docker host, is able to successfully sync the Plex server.

- Installed Preroll Plus using a stack in Portainer.
- Webhook set up in Plex settings.
- In Plex settings, Webhooks and Push Notifications are both checked/enabled.
- This is the only webhook configured in Plex.
- Successfully authenticated with Plex in Preroll Plus gui.

![image](https://github.com/user-attachments/assets/886fa0a5-9af0-4411-b6b4-b46779f16b5a)

```
> start
> nodemon ./bin/www
[nodemon] 2.0.22
[nodemon] to restart at any time, enter `rs`
[nodemon] watching path(s): *.*
[nodemon] watching extensions: js,mjs,json
[nodemon] starting `node ./bin/www`
[2024-09-12 10:11:18] [INFO]  Config file updated to UID: 1000 GID: 1000
[2024-09-12 10:11:18] [INFO]  Settings file read
[9/12/2024 10:16:21] [INFO]  GET / - Status: 200
[9/12/2024 10:16:21] [INFO]  GET /static/css/main.6fa433e6.css - Status: 200
[9/12/2024 10:16:21] [INFO]  GET /static/js/main.c15f3406.js - Status: 200
[2024-09-12 10:16:21] [INFO]  Settings file read
[9/12/2024 10:16:21] [INFO]  GET /backend/load - Status: 200
[9/12/2024 10:16:21] [INFO]  GET /static/media/loading-gif.803f24369a085e35482e.gif - Status: 200
[9/12/2024 10:16:21] [INFO]  GET /apple-touch-icon.png - Status: 200
[9/12/2024 10:16:21] [INFO]  GET /favicon.ico - Status: 200
[9/12/2024 10:16:22] [INFO]  GET /static/media/box-arrow-in-left.031730b524a3003d5ba59a56fceba711.svg - Status: 200
[2024-09-12 10:16:49] [INFO]  Retrieving thumbnail from Plex account
[9/12/2024 10:16:49] [INFO]  POST /backend/thumb - Status: 200
[2024-09-12 10:16:49] [INFO]  Settings file saved
[9/12/2024 10:16:49] [INFO]  POST /backend/save - Status: 200
[9/12/2024 10:16:49] [INFO]  GET /login - Status: 200
[9/12/2024 10:16:49] [INFO]  GET /static/js/main.c15f3406.js - Status: 304
[9/12/2024 10:16:49] [INFO]  GET /static/css/main.6fa433e6.css - Status: 304
[9/12/2024 10:16:49] [INFO]  GET /static/media/loading-gif.803f24369a085e35482e.gif - Status: 304
[2024-09-12 10:16:49] [INFO]  Settings file read
[9/12/2024 10:16:49] [INFO]  GET /backend/load - Status: 200
[9/12/2024 10:16:49] [INFO]  GET /static/media/info-circle.b8b4fd4841fa78267fd939ea96bd252f.svg - Status: 200
[9/12/2024 10:16:49] [INFO]  GET /static/media/arrow-repeat.a5aaafc7324141569da4e6f2e5673eec.svg - Status: 200
[2024-09-12 10:17:38] [INFO]  Retrieving Plex Servers
[2024-09-12 10:17:38] [ERROR] Issue with connection to Plex Server:  Invalid URL
[2024-09-12 10:17:48] [ERROR] Issue with connection to Plex Server:  timeout of 10000ms exceeded
[9/12/2024 10:17:48] [INFO]  POST /backend/settings - Status: 401
[2024-09-12 10:17:52] [INFO]  Retrieving Plex Servers
[2024-09-12 10:17:52] [ERROR] Issue with connection to Plex Server:  Invalid URL
[2024-09-12 10:18:02] [ERROR] Issue with connection to Plex Server:  timeout of 10000ms exceeded
[9/12/2024 10:18:02] [INFO]  POST /backend/settings - Status: 401
[2024-09-12 10:18:05] [INFO]  Retrieving Plex Servers
[2024-09-12 10:18:05] [ERROR] Issue with connection to Plex Server:  Invalid URL
[2024-09-12 10:18:15] [ERROR] Issue with connection to Plex Server:  timeout of 10000ms exceeded
[9/12/2024 10:18:15] [INFO]  POST /backend/settings - Status: 401
[9/12/2024 10:20:10] [INFO]  GET /static/media/box-arrow-right.ac16fec15d0de6697376f367316c8542.svg - Status: 200
[2024-09-12 10:21:10] [INFO]  Retrieving Plex Servers
[2024-09-12 10:21:10] [ERROR] Issue with connection to Plex Server:  Invalid URL
[2024-09-12 10:21:20] [ERROR] Issue with connection to Plex Server:  timeout of 10000ms exceeded
[9/12/2024 10:21:20] [INFO]  POST /backend/settings - Status: 401
```

**Comment by chadwpalm** [comment]
I added some additional debug output in that section of code when the arrow circle is pressed. This should hopefully tell us what might be going wrong here.

I pushed an update to the develop branch that has those changes. If you could pull that and recreate the container (the app will make you log in again when there's a version change), you should be able to confirm it is build number 9 from the About window on the top right menu pull-down.

Then try the arrow circle again and look at the logs. What the app is doing is retrieving a list of servers claimed by your account, then using the local IP address to create the URL that grabs the cert from your server(s) (this is for secure connections) and then generates the list to display in the server pull-down menu in the GUI.

The updated debug should show how many servers it found and then go through each and display all the server information. I've also added output of the URL string that it creates from that to grab the cert. I believe this is where the problem lies, so that's why I want to output all the server info.

Be aware if you paste the logs here again, it will show your public facing IP and port in the debug (part of what I added), so you might want to be careful to block that out.

We'll go from there.

Chad

**Comment by imsorrywhat** [comment]
Updated to build 9. External IPs sanitized and left out some of the irrelevant lines; if you need everything, let me know.

Plex is running as an app in TrueNAS Scale. Looks like it's using the list of IPs returned in Local Address to form the URL which explains the Invalid URL. Not entirely sure why the 172 IPs are being included. I'm guessing something to do with kubernetes network.

The 2nd server, plexbox2, is a shared external server. No issues with the cert URL but the timeout is expected since it's trying to reach an IP that doesn't exist in my network. Should external servers be ignored since I would have no control over their preroll configuration? edit: not sure if it's possible to differentiate though.

Thanks!

```
...
[2024-09-12 13:08:22] [INFO]  Version updated from 0.1.0 build 8 branch dev to 0.1.0 build 9 branch dev
[2024-09-12 13:08:22] [INFO]  Config file updated to UID: 1000 GID: 1000
[2024-09-12 13:08:22] [INFO]  Settings file read
[9/12/2024 13:08:41] [INFO]  GET /settings - Status: 200
...
[2024-09-12 13:09:21] [INFO]  Retrieving Plex Servers
[2024-09-12 13:09:21] [DEBUG] Found 2 Plex servers
[2024-09-12 13:09:21] [DEBUG] Server 1 info:
[2024-09-12 13:09:21] [DEBUG] Name: plexbox
[2024-09-12 13:09:21] [DEBUG] External Address: 1.1.1.1
[2024-09-12 13:09:21] [DEBUG] External Port: 32400
[2024-09-12 13:09:21] [DEBUG] Local Address: 192.168.2.128,172.16.0.1,172.17.0.1,172.17.0.10,172.17.252.23,172.17.160.226,172.17.87.161,172.17.74.155
[2024-09-12 13:09:21] [DEBUG] Local Port: 32400
[2024-09-12 13:09:21] [DEBUG] Server 2 info:
[2024-09-12 13:09:21] [DEBUG] Name: plexbox2
[2024-09-12 13:09:21] [DEBUG] External Address: 1.1.1.1
[2024-09-12 13:09:21] [DEBUG] External Port: 32400
[2024-09-12 13:09:21] [DEBUG] Local Address: 192.168.50.163
[2024-09-12 13:09:21] [DEBUG] Local Port: 32400
[2024-09-12 13:09:21] [DEBUG] Retrieving Cert from URL: http://192.168.2.128,172.16.0.1,172.17.0.1,172.17.0.10,172.17.252.23,172.17.160.226,172.17.87.161,172.17.74.155:32400/:/prefs
[2024-09-12 13:09:21] [ERROR] Issue with connection to Plex Server while retrieving cert:  Invalid URL
[2024-09-12 13:09:21] [DEBUG] Retrieving Cert from URL: http://192.168.50.163:32400/:/prefs
[2024-09-12 13:09:31] [ERROR] Issue with connection to Plex Server while retrieving cert:  timeout of 10000ms exceeded
[9/12/2024 13:09:31] [INFO]  POST /backend/settings - Status: 401
```

**Comment by chadwpalm** [comment]
This is good info, thank you. That's why I'm beta testing because there can be scenarios that I can't see.

I didn't think about the fact that it would show servers that you are remotely added to. Thought it was just ones you owned. Thankfully, it looks like the XML that comes back during the server retrieval does have a flag called "owned". I can filter out those since they won't be accessible like you said.

The IP thing is interesting. Didn't expect it would/could return a comma-delimited list of IPs.  I can add a parser to list them all out.....but it needs the correct one for accessing the server. In your case it's the first one, correct? Don't know if I can guarentee it will always be the first one.....but.....I can also have it go through each local address until it finds one the server responds with, or just have it fall through when it times out.

In the meantime, have you tried to just manually input the correct IP and port to at least see if the rest of the app works?

**Comment by chadwpalm** [comment]
Also, when you do the same thing in Overseerr (click the arrow circle) does it only show your correct local IP or does it also show all of the 172 addresses? I also assume Overseerr filters out unowned servers, but I'm curious if you see those too.

Thanks.

**Comment by imsorrywhat** [comment]
Happy to help!

Yes, it's the 192 for mine. Probing the IPs for a response if comma delimited makes sense to me. I would guess that Plex installs in TrueNAS Scale are probably in the minority. And with the next major release, Kubernetes is being replaced by Docker. Perhaps it won't be an issue then.

As for Overseerr, forgot to mention that earlier. It filters out unowned servers and does not show them. It does show the local, public, and 172 IPs but the 172 are not available to select, they show as Timed Out / Conn Refused.

Just tried adding IP manually and it saved in GUI and logs also show a successful save so I'm assuming manual IP works. Going to work on finishing the setup.

```
[2024-09-12 15:47:44] [INFO]  Settings file saved
[9/12/2024 15:47:44] [INFO]  POST /backend/save - Status: 200
```

**Comment by chadwpalm** [comment]
Thanks again....lots of useful info.

So, Preroll Plus accesses the Plex server to update the pre-roll string on three occasions:

1. When you create or edit a Sequence
2. When you start playing a movie after pre-rolls are finished
3. Every night at midnight as it checks the current date against all the scheduled sequences.

All three of these scenarios look at all the Sequences and fills in the string with the one that is in its schedule, or falls back to a non-scheduled Sequence.

If you want to do a quick test, create a Sequence without a schedule and see if it updates on the Plex server. If not, check the logs and see if there was an error.

Also, I'd like to know if in the Buckets tab you actually see your pre-roll media in the file list.

Thanks.

In the meantime, I'll make adjustments to the code to fix the server detection issues.

**Comment by imsorrywhat** [comment]
We're in bidniz. :)

I confirmed 1 and 2 work as you stated. I did notice that an extra / is added in the string. It doesn't seem to impact functionality as the preroll file does play and does update.

<img width="429" alt="Screenshot 2024-09-12 at 7 02 49 PM" src="https://github.com/user-attachments/assets/8d1bf1a8-d592-4977-8a73-838e2febe1da">  
 
A question about Sequences. So I first created a scheduled one, from 9/1-9/30 and called it Christmas. This worked as expected. I then created a new Sequence for another holiday, no schedule. Once that was saved, it updated Plex and overrode the Christmas string. Shouldn't it have not updated since the existing one is within schedule and new one has no schedule?

In Plex, the original preroll path was ```/media/Plex Prerolls/```. Both that one and the path I used in docker compose,
``` /mnt/Plex Media/Plex Prerolls```, seem to work just fine. Not an issue, just thought I'd mention that.

I do see the preroll file list in the Buckets tab.

![image](https://github.com/user-attachments/assets/9326f53a-0ea0-4ceb-b65d-0ba1ded0c93f)

I've come across one other bug that I'd say is minor. 

- create a new Sequence
- without selecting any Buckets, click the Add button.
- the page goes blank.
- refreshing the page fixes it but the Sequence is closed out. 

I'm on latest Firefox. Let me know if there's any data you'd like collected or if you'd prefer I log it separately.

**Reply by chadwpalm** [reply]
> I did notice that an extra / is added in the string. It doesn't seem to impact functionality as the preroll file does play and does update.

Weird...mine doesn't do that. Did you add a slash at the end of the entry "Plex location of preroll media" in settings, or add a slash at the end of the volume mount for the container? Perhaps it's another weird artifact from Kubernetes. I guess as long as it is not affecting the functionality it's okay, but still weird.

> So I first created a scheduled one, from 9/1-9/30 and called it Christmas. This worked as expected. I then created a new Sequence for another holiday, no schedule. Once that was saved, it updated Plex and overrode the Christmas string. Shouldn't it have not updated since the existing one is within schedule and new one has no schedule?

I just tried to replicate what you did, but I didn't have an issue. It will perform a string update when you create a new sequence, but it should have updated the string for the sequence within the schedule (Christmas). If there is more than one file in a bucket within that sequence it will pick a new random file for that bucket (as if you had played the movie in Plex). Are you sure it didn't just change to another file in the bucket...or that you don't have the same file in two buckets...or that you are maybe using the same bucket in both sequences? Perhaps make buckets with a single file for each sequence to make sure it always chooses that one file....if you haven't already tried that.

I did notice that I don't have the string update on a sequence deletion and need to add that in.

> I've come across one other bug that I'd say is minor.
>
> - create a new Sequence
> - without selecting any Buckets, click the Add button.
> - the page goes blank.
> - refreshing the page fixes it but the Sequence is closed out.
>
>I'm on latest Firefox. Let me know if there's any data you'd like collected or if you'd prefer I log it separately.

For me in both Chrome (Version 128.0.6613.119) and Firefox (Version 129.0.2) (Both in Ubuntu) if I leave the title blank or fail to add a bucket, I get this:

![image](https://github.com/user-attachments/assets/2e2b5a41-e0aa-4efa-b906-6b20ab230fae)
(And just realized it should say sequence, not bucket)

Are you in Linux or Windows and which version of FF are you using so I can try and replicate? If you hit F12 it will open debug and go to console tab and see is there's an error message there when you do it.

Chad

**Comment by imsorrywhat** [comment]
No extra / but no worries. Not breaking anything.
![image](https://github.com/user-attachments/assets/2f191e97-bd58-4196-b46b-764761540488)
![image](https://github.com/user-attachments/assets/e0e346fd-133b-4009-a135-4ec740f1b1de)

Yeah, disregard the Sequence question. It does update to another preroll in the same bucket. I got my file names mixed up. User error. :P

I'm on 

- MacOS Sonoma 14.6.1
- Firefox v130.0
- Same happens in Safari v17.6.

Left arrow/Add button:
![image](https://github.com/user-attachments/assets/325625d9-446e-4602-9a79-df120f5829ab)

After clicking it:
![image](https://github.com/user-attachments/assets/6377feb7-2e41-462f-8411-15e451db0085)

**Comment by chadwpalm** [comment]
Ok, I can replicate the problem. For some reason I was thinking you meant the "Save" button rather than the "Add" button. My fault.

I had some family stuff to deal with yesterday, so I'll be working on all of these issues today and (fingers crossed) should have the fixes done today.

**Comment by imsorrywhat** [comment]
No worries!

I set up more buckets and have a few questions:

- the file list is sorted A-Z, would it be possible to also sort A-Z when adding files to the bucket instead of order they're added? would make it easier to track what's been added when working with a large preroll library. i have 111 and some file names can be similar to each other.
- can the page banner be anchored (not sure correct term) so that it doesn't disappear when scrolling? mostly happens when editing/creating a new bucket.
- can the "adding" flow be the same for Sequences and Buckets? former has the Left Arrow, latter has the Add buttons at the top.
- can you share an icon file that can be used with dashboards? (i use homepage)
- can you add a visual indicator to the bucket that is being edited? for example, change the color of the outline to red.
- probably a much bigger task; would it be possible to embed a player in the Buckets page to preview prerolls?
- probably more of a corner case but if the file name is changed at the source, and the bucket is not updated with new file name, preroll won't play which is expected. can some sort of file check be added so that Preroll alerts if file "goes missing"?

Sorry, that ended up being more than a few! I'm really enjoying this though. Have always wanted to be able to do exactly what Preroll does so Thank You!

**Reply by chadwpalm** [reply]
> the file list is sorted A-Z, would it be possible to also sort A-Z when adding files to the bucket instead of order they're added? would make it easier to track what's been added when working with a large preroll library. i have 111 and some file names can be similar to each other.

This should be easy enough to implement and a good idea.

> can the page banner be anchored (not sure correct term) so that it doesn't disappear when scrolling? mostly happens when editing/creating a new bucket.

I don't think this would be a problem to do, and again, a good idea....one I may cross implement into Lumunarr.

> can the "adding" flow be the same for Sequences and Buckets? former has the Left Arrow, latter has the Add buttons at the top.

Yup. This was the case of working on one for a long while and then moving to the other and not really seeing the forest for the trees lol.

> can you share an icon file that can be used with dashboards? (i use homepage)

https://drive.google.com/file/d/1IbUgZJFqL7BCixDMvZXHHTFOVAPdwrMK/view?usp=sharing

That should have the favicon pack, a large .png with transparent background, and the original photoshop .psd file. Have fun!

> can you add a visual indicator to the bucket that is being edited? for example, change the color of the outline to red.

Love it! Good idea! Again, something I can cross-implement to Lumunarr.

> probably a much bigger task; would it be possible to embed a player in the Buckets page to preview prerolls?

Yeah.....I'm still an amateur when it comes to web development, so I wouldn't get my hopes up, but this is an open source project, so maybe someone who sees this can tackle it, but maybe down the road when I have time I can at least look into it. Maybe not as hard as I think it might be. :)

> probably more of a corner case but if the file name is changed at the source, and the bucket is not updated with new file name, preroll won't play which is expected. can some sort of file check be added so that Preroll alerts if file "goes missing"?

At the least, I can maybe have it check if the files still exist when you open up the bucket for editing and check it against the files in the current directory. Not sure if I could have any sort of constant scanning of the directory though. The frontend UI only contacts the backend (which is what does the file handling) when an action on the frontend is taken (like a button press). I'll see what I can do though.

> Sorry, that ended up being more than a few! I'm really enjoying this though. Have always wanted to be able to do exactly what Preroll does so Thank You!

No problem! Let me take care of the bug fixes first and I can start tackling these as feature requests over the weekend.

I came up with this idea several months ago and because I'm in-between jobs right now so I have the free time to do so.

**Reply by chadwpalm** [reply]
> * the file list is sorted A-Z, would it be possible to also sort A-Z when adding files to the bucket instead of order they're added? would make it easier to track what's been added when working with a large preroll library. i have 111 and some file names can be similar to each other.
> * can the page banner be anchored (not sure correct term) so that it doesn't disappear when scrolling? mostly happens when editing/creating a new bucket.
> * can the "adding" flow be the same for Sequences and Buckets? former has the Left Arrow, latter has the Add buttons at the top.
> * can you share an icon file that can be used with dashboards? (i use homepage)
> * can you add a visual indicator to the bucket that is being edited? for example, change the color of the outline to red.
> * probably a much bigger task; would it be possible to embed a player in the Buckets page to preview prerolls?
> * probably more of a corner case but if the file name is changed at the source, and the bucket is not updated with new file name, preroll won't play which is expected. can some sort of file check be added so that Preroll alerts if file "goes missing"?

I just pushed 0.1.2 which in which I implemented everything in this list except the last bullet. I'm still thinking about how I want to deal with backend file/directory monitoring before moving forward so I am not going back and forth on things. Please feel free to join in the discussion about it in the other thread [https://github.com/chadwpalm/PrerollPlus/discussions/2](https://github.com/chadwpalm/PrerollPlus/discussions/2)

For the video preview, it will play all of the videos sequentially that are selected from the top down. You can stop playback at anytime with the stop button and you can interrupt playback with a new set of videos by clicking Preview again. Let me know how it works out on your end.

**Reply by imsorrywhat** [reply]
Everything checks out! Absolutely loving the preview option. Nice job adding that in! I'll check out the other thread, see what's happening over there. Appreciate your work on this!

**Comment by imsorrywhat** [comment]
Awesome, thanks for the icon files! Dashboard looks better.

Forgot to include Sequences in the visual indicator request.

Also, in the [User Guide](https://github.com/chadwpalm/PrerollPlus/wiki/User-Guide), I'm guessing this sentence should say Preroll Plus instead of Lumunarr?

- The port number will depend on whether **_Lumunarr_** is running natively or within a Docker container.

**Reply by chadwpalm** [reply]
> Awesome, thanks for the icon files! Dashboard looks better.

No prob.

> Forgot to include Sequences in the visual indicator request.

Gotcha.

> Also, in the [User Guide](https://github.com/chadwpalm/PrerollPlus/wiki/User-Guide), I'm guessing this sentence should say Preroll Plus instead of Lumunarr?
> * The port number will depend on whether **_Lumunarr_** is running natively or within a Docker container.

Fixed.

I just pushed out 0.1.1.10 to the develop branch. I think I got what we've talked about fixed, but the  two things I'd like you to specifically look at are the server retrieval behavior and if I fixed the double slash issue (and not made anything worse).

For server retrieval, I should have filtered out non-owned servers and it will gray out IP's where it couldn't contact a server.

All changes are in the history.md file.

**Comment by imsorrywhat** [comment]
Both issues fixed.

Plex server retrieval works. No remote servers listed, unreachable IPs grayed out.
![IMG_0348](https://github.com/user-attachments/assets/3844f18e-2498-4395-bf00-108c01f43978)

No more double slash.
![image](https://github.com/user-attachments/assets/ede27098-dc78-4c72-975e-951c8d1e65f0)
